File: app/src/main/java/com/botijasoftware/utils/collision/CollisionHollowSphere.java
package com.botijasoftware.utils.collision;

import com.botijasoftware.utils.Vector3;

public class CollisionHollowSphere extends CollisionVolume {
	private Sphere outsphere;
	private Sphere insphere;
	private float holeangle;
	private float halfholeangle;
	private float angle;
	
	
	public CollisionHollowSphere(Vector3 center, float outradius, float inradius, float holeangle) {
		outsphere = new Sphere(center, outradius);
		insphere = new Sphere(center, inradius);
		this.holeangle = holeangle;
		this.halfholeangle = this.holeangle/2;
	}
	
	private boolean inHole(Vector3 v) {
		float a = (float)Math.atan2(-(v.Y-outsphere.Y), v.X-outsphere.X);
		return (a < angle+halfholeangle) && (a > angle-halfholeangle); 
	}
	
	public boolean collides(CollisionVolume c) {
		return (c.collides(outsphere) && (!c.collides(insphere)) && !inHole(c.getPosition())); 
	}
	
	public boolean collides(Sphere s) {
		return (outsphere.collides(s) && (!insphere.collides(s) && !inHole(s.getCenter()))) ;
	}
	
	public boolean collides(Vector3 v) {
		return (outsphere.collides(v) && (!insphere.collides(v) && !inHole(v)));
	}
	
	public Vector3 getPosition() {
		return insphere.getCenter();
	}
	
	public void setPosition(Vector3 v) {
		outsphere.setCenter(v);
		insphere.setCenter(v);
	}
	
	public void setOutRadius(float r) {
		outsphere.setRadius(r);
	}
	
	public void setInRadius(float r) {
		insphere.setRadius(r);
	}
	
	public float getOutRadius() {
		return outsphere.radius;
	}
	
	public float getInRadius() {
		return insphere.radius;
	}
	
	public void rotate(float angle) {
		this.angle = angle;
	}
	
}

File: app/src/main/java/com/botijasoftware/utils/collision/CollisionHollowSphere.kt
package com.botijasoftware.utils.collision

import com.botijasoftware.utils.Vector3

class CollisionHollowSphere(center: Vector3, outradius: Float, inradius: Float, private val holeangle: Float) : CollisionVolume() {
    private val outsphere: Sphere
    private val insphere: Sphere
    private val halfholeangle: Float
    private var angle: Float = 0.toFloat()


    init {
        outsphere = Sphere(center, outradius)
        insphere = Sphere(center, inradius)
        this.halfholeangle = this.holeangle / 2
    }

    private fun inHole(v: Vector3): Boolean {
        val a = Math.atan2((-(v.Y - outsphere.Y)).toDouble(), (v.X - outsphere.X).toDouble()).toFloat()
        return a < angle + halfholeangle && a > angle - halfholeangle
    }

    override fun collides(c: CollisionVolume): Boolean {
        return c.collides(outsphere) && !c.collides(insphere) && !inHole(c.position)
    }

    override fun collides(s: Sphere): Boolean {
        return outsphere.collides(s) && !insphere.collides(s) && !inHole(s.center)
    }

    override fun collides(v: Vector3): Boolean {
        return outsphere.collides(v) && !insphere.collides(v) && !inHole(v)
    }

    override var position: Vector3
        get() = insphere.center
        set(v) {
            outsphere.center = v
            insphere.center = v
        }

    var outRadius: Float
        get() = outsphere.radius
        set(r) { outsphere.radius = r }

    var inRadius: Float
        get() = insphere.radius
        set(r) { insphere.radius = r }

    fun rotate(angle: Float) {
        this.angle = angle
    }

}